File: Serveur/Socket/RicochetSocketsServer/Socket/Packet/CustomPacket/DataPacket.cpp
#include "DataPacket.h"

#include <cstring>

DataPacket::DataPacket() : Packet(), isOwner(true) {
    this->type = Type::DATA;
    this->data = nullptr;
    this->dataSize = 0;
}

DataPacket::DataPacket(const char* data, size_t size, bool isOwner) : Packet(), isOwner(isOwner){
    this->type = Type::DATA;
    this->data = new char[size];
    memcpy(this->data, data, size);
    this->dataSize = size;
}

DataPacket::~DataPacket() {
    if(isOwner) delete[] data;
}

DataPacket* DataPacket::deserialize(const char* serialized) {
    Type type = deserializeType(serialized);
    if(type != Type::DATA) return nullptr;

    auto* packet = new DataPacket();
    size_t size = *reinterpret_cast<const size_t*>(serialized);
    packet->dataSize = size - sizeof(size_t) - sizeof(Type);
    packet->data = new char[packet->dataSize];
    memcpy(packet->data, serialized + sizeof(size_t) + sizeof(Type), packet->dataSize);
    packet->isOwner = true;

    return packet;
}

File: Serveur/Socket/JavaSocketClientExample/untitled/src/RicochetSocket/Packet/CustomPacket/DataPacket.java
package RicochetSocket.Packet.CustomPacket;

import RicochetSocket.Packet.Packet;

/**
 * Data packet class, used to represent a data packet
 */
public class DataPacket extends Packet {
    /**
     * Constructor
     */
    public DataPacket() {
        super();
        this.type = Type.DATA;
    }

    /**
     * Constructor
     * @param data Data
     */
    public DataPacket(byte[] data) {
        this.type = Type.DATA;
        this.data = data;
    }

    /**
     * Deserialize the packet
     * @param data Serialized packet
     */
    public static DataPacket deserialize(byte[] data) {
        Type type = deserializeType(data);
        if(type!=Type.DATA) return null;

        DataPacket dataPacket = new DataPacket();
        long size = 0;
        for(int i = 0; i < 8; i++) {
            size |= (long) (data[i] & 0xFF) << (i * 8);
        }
        long dataSize = size - dataPacket.getHeaderSerializationSize();
        dataPacket.data = new byte[(int) dataSize];
        System.arraycopy(data, 9, dataPacket.data, 0, (int) dataSize);
        return dataPacket;
    }
}